### DISCUSSION TITLE
gen_server:call vs rpc:call from erlang to ergo

### DISCUSSION THREAD

**Discussion by xiaodoudou12** [ergo-services/ergo#146] [Q&A] [answered]
```
type demo struct {
        gen.Server
}
func (d *demo) HandleCast(process *gen.ServerProcess, message etf.Term) gen.ServerStatus {
        return gen.ServerStatusOK
}
func (d *demo) HandleCall(process *gen.ServerProcess, from gen.ServerFrom, message etf.Term) (etf.Term, gen.ServerStatus) {
    time.Sleep(time.Second)
    return message, gen.ServerStatusOK
}
var (
        ServerName string
        NodeName   string
        Cookie     string
)
func init() {
        flag.StringVar(&ServerName, "server", "example", "server process name")
        flag.StringVar(&NodeName, "name", "test_client_39@10.2.18.22", "node name")
        flag.StringVar(&Cookie, "cookie", "test_dev_39", "cookie for interaction with erlang cluster")
}
func main() {
        flag.Parse()
        node, err := ergo.StartNode(NodeName, Cookie, node.Options{})
        if err != nil {
                panic(err)
        }
        _, err = node.Spawn(ServerName, gen.ProcessOptions{}, &demo{})
        if err != nil {
                panic(err)
        }
        .....
        node.Wait()
```
I simulate a function need one second in ergo.
it will be as a server .
```
#!/usr/bin/env escript
-mode(compile).
main(_) ->
   SelfNode = list_to_atom("erl-test_client_39@10.2.18.22"),%atom, self node name
   Cook = list_to_atom("test_dev_39"),% atom,cookie
   net_kernel:start([SelfNode, longnames]),%
   erlang:set_cookie(node(), Cook),
   test_fun(0),
   [begin
      spawn(fun() -> test_fun(Index) end)
    end || Index <- lists:seq(1, 5)],
   timer:sleep(7000),
   ok.

timenow() ->
  {A, B, _} = erlang:timestamp(),
  A * 1000 * 1000 + B.

test_fun(Index) ->
  StartTime =timenow(),
  io:format("start_gen_call [~p] time:~p ~n", [Index, StartTime]),
  gen_server:call({example,'test_client_39@10.2.18.22'}, {echo, 1,2,3}),
  EndTime = timenow(),
  io:format("finish_gen_call [~p] time:~p usetime:~p ~n", [Index, EndTime, EndTime-StartTime]),
  ok.
```
the escript will be a client gen_server:call from erlang to ergo.
```
start_gen_call [0] time:1678417909 
finish_gen_call [0] time:1678417910 usetime:1 
start_gen_call [1] time:1678417910 
start_gen_call [2] time:1678417910 
start_gen_call [3] time:1678417910 
start_gen_call [4] time:1678417910 
start_gen_call [5] time:1678417910 
finish_gen_call [1] time:1678417911 usetime:1 
finish_gen_call [3] time:1678417912 usetime:2 
finish_gen_call [5] time:1678417913 usetime:3 
finish_gen_call [2] time:1678417914 usetime:4 
```
Because gen_server process command by sequence,the gen_call from client will run one by one, so the QPS is very low.
gen_server:call 1 use 1 second
gen_server:call 2 use 2 second
gen_server:call 3 use 3 second
....
now i only can start example_1,example_2,....example_10 for 10 gen_server in ergo.
i can rand from example_1,example_2,....example_10 for call in erlang.

if i change gen_server:call to rpc:call
```
rpc:call(list_to_atom("erl-test_39@10.2.18.22"), test_rpcserver, test_call, [Index]),

// this is example test_call
test_call(Index) ->
  io:format("start test call ~p~n", [Index]),
  timer:sleep(1000),
  io:format("finish test call ~p~n", [Index]),
  ok. 
```
in this way rpc:call from erlang to elang.
rpc:call 1 use 1 second.
rpc:call 2 use 1 second too.
rpc:call 3 use 1 second too.
....
so i think gen_server:call from erlang to ergo is slow in sometimes,we also need make rpc:call run in spawn child like erlang erpc.

[https://www.erlang.org/blog/otp-23-highlights/#distributed-spawn-and-the-new-erpc-module](url)

**Comment by halturin** [comment, marked as answer]
this output...
```
finish_gen_call [1] time:1678417911 usetime:1 
finish_gen_call [3] time:1678417912 usetime:2 
finish_gen_call [5] time:1678417913 usetime:3 
finish_gen_call [2] time:1678417914 usetime:4 
```
says you handle your requests using the same processes, which is sleeping for a second handling every message.

Erlang' ERPC handles requests spawning a new process for every request.

**Reply by xiaodoudou12** [reply]
I only want make ergo server comunicate with erlang client.
i can only use gen_server:call in erlang to ergo.like the example ergo make a GenServer receive the message from erlang.
it work ok, but the GenServer is single point,
I want know how can i handle my request in different processes in ergo.
i will learn how to use RemoteSpawn in tests/node_test.go also.

**Reply by halturin** [reply]
I guess you are looking for a kind of pool of workers, so you may try `gen.Pool` for that (it was introduced in the latest release). Here is the example https://github.com/ergo-services/examples/tree/master/genpool

**Reply by xiaodoudou12** [reply]
oh, this is i wanted, thank you very much.

**Comment by halturin** [comment]
Basically, ERPC is an abstraction on top of the RemoteSpawn. Ergo supports this kind of request, so you may want to try it. See TestNodeRemoteSpawn(..) in tests/node_test.go as an example.
